### PR TITLE
Adding partial instance

### PR DESCRIPTION
**Pull request by vmontagh**

### CODE REVIEW

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    public final Expr fact;
+    
+    public Bounds(String label){
+        super(Pos.UNKNOWN, null);
+        this.label = label;
+        scope = new ArrayList<CommandScope>();
+        funcs = new ArrayList<ExprVar>();
+        fact = null;
+
+
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: too many empty lines

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import edu.mit.csail.sdg.alloy4.ConstList;
+import edu.mit.csail.sdg.alloy4.Err;
+import edu.mit.csail.sdg.alloy4.ErrorWarning;
+import edu.mit.csail.sdg.alloy4.Pos;
+
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: let's add some doc comments for public classes/fields

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        this.label = label;
+        scope = new ArrayList<CommandScope>();
+        funcs = new ArrayList<ExprVar>();
+        fact = null;
+
+        
+    }
+
+    public Bounds(Pos pos, String label, List<CommandScope> list){
+        super( Pos.UNKNOWN,Type.FORMULA);
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: let's try to keep spacing consistent.  For example, here you used `super( Pos...)` and above you used `super(Pos...)`

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    public Bounds(Pos pos, String label, List<CommandScope> list, Expr fact){
+        super( Pos.UNKNOWN,Type.FORMULA);
+        this.label = label;
+        this.scope = new ArrayList<CommandScope>(list);
+        funcs = new ArrayList<ExprVar>();
+        this.fact = fact;
+    }
+
+
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: too many new lines

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    
+    public Bounds(Pos pos, String label, List<CommandScope> list, List<ExprVar> funcs){
+        super( Pos.UNKNOWN,Type.FORMULA);
+        this.label = label;
+        this.scope = new ArrayList<CommandScope>(list);
+        this.funcs = new ArrayList<ExprVar>(funcs);
+        fact = null;
+    }
+
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: too many empty lines

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    
+    @Override
+    public void toString(StringBuilder out, int indent) {
+        // TODO Auto-generated method stub
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: let's not leave such TODO's

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        return null;
+    }
+
+    @Override
+    public String toString() {
+        return "Bounds [label=" + label + ", scope=" + scope + ", funcs="
+                + funcs + ", fact=" + fact + "]";
+    }
+
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: empty line before closing brace

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java [hunk 8]
```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
...
-            throw new ErrorSyntax(pos, "Sig " + sig + " cannot have a negative starting scope (" + startingScope + ")");
-        if (endingScope < 0)
-            throw new ErrorSyntax(pos, "Sig " + sig + " cannot have a negative ending scope (" + endingScope + ")");
-        if (endingScope < startingScope)
-            throw new ErrorSyntax(pos, "Sig " + sig + " cannot have an ending scope (" + endingScope + ") smaller than its starting scope (" + startingScope + ")");
-        if (startingScope == endingScope)
-            increment = 1;
-        if (increment < 1)
-            throw new ErrorSyntax(pos, "Sig " + sig + "'s increment value cannot be " + increment + ".\nThe increment must be 1 or greater.");
+        this(pos, sig, isExact, startingScope, endingScope,increment,false);
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: needs a space after a comma

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java [hunk 9]
```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
...
+    }
+
+    /**
+     * It is similar the previous constructor, but sets the sparse boolean value to see whether integer values in inst-block should be included in the
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: "similar **to** the previous constructor"

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java [hunk 10]
```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
...
+     * atoms or not. 
+     * @param pos
+     * @param sig
+     * @param isExact
+     * @param startingScope
+     * @param endingScope
+     * @param increment
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: either add doc comments for these parameters or delete `@param` (just don't leave them empty)

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java [hunk 11]
```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
...
+    }
+
+    /**
+     * Create a commandscope for a sig. 
+     * If the atoms list is empty, it no sig should be created.  
+     * If not isExact and LIndex = 0 and atoms.size() > 0 then hasUpper and notLower
+     * If not isExact and LIndex > 0 and atoms.size() = LIndex then HasLower
+     * If not isExact and LIndex > 0 and atoms.size() > LIndex the hasLower and hasUpper, means a range.
+     * If isExact then LIndex = atoms.size() = LIndex
+     * @param pos
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: same here

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java [hunk 12]
```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
...
+        if (startingScope == endingScope) increment = 1;
+        if (increment < 1) throw new ErrorSyntax(pos, "Sig "+sig+"'s increment value cannot be "+increment+".\nThe increment must be 1 or greater.");
         this.pos = pos;
         this.sig = sig;
         this.isExact = isExact;
-        this.startingScope = startingScope;
-        this.endingScope = endingScope;
+        this.hasUpper = !isExact & LIndex < atoms.size();
+        this.hasLower = !isExact & LIndex>0;
+        this.pAtomsLowerLastIndex=LIndex;
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: needs spaces around `=`

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java [hunk 13]
```diff
@@ -72,47 +94,174 @@ public CommandScope(Sig sig, boolean isExact, int scope) throws ErrorSyntax {
...
         this.increment = increment;
+        this.pFields = new ArrayList<List<Expr>>();
+        this.isSparse = isSparse;
+        boolean isRelation = false;
+
+        List<List<Expr>> tmpPField = new ArrayList<List<Expr>>();
+        if(atoms != null  && atoms.size() > 0){
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: needs space between `if` and and open paren; same for `for` below

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompSym.java [hunk 14]
```diff
@@ -1,7 +1,7 @@
 
 //----------------------------------------------------
 // The following code was generated by CUP v0.11a beta 20060608
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompSym.java]: why do we have this file checked in...

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java [hunk 15]
```diff
@@ -313,24 +323,31 @@
...
-        if (max >= min)
-            for (int i = min; i <= max; i++) { // Safe since we know 1 <=
-                                              // bitwidth <= 30
-                Tuple ii = factory.tuple("" + i);
+        if (max >= min){
+            ConcurrentSkipListSet<Integer> ints = new ConcurrentSkipListSet<Integer>();
+            for(String a: atoms){
+                try{
+                    ints.add(Integer.valueOf(a));
+                }catch(NumberFormatException e){}
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java]: Catching `NumberFormatException` and ignoring it does not seem safe.  If this exception can happen due to a user error, I'd rather it blows up all the way to the top

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java [hunk 16]
```diff
@@ -313,24 +323,31 @@
         TupleSet stringBounds = factory.noneOf(1);
         final TupleSet next = factory.noneOf(2);
         int min = min(), max = max();
-        if (max >= min)
-            for (int i = min; i <= max; i++) { // Safe since we know 1 <=
-                                              // bitwidth <= 30
-                Tuple ii = factory.tuple("" + i);
+        if (max >= min){
+            ConcurrentSkipListSet<Integer> ints = new ConcurrentSkipListSet<Integer>();
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java]: Why `ConcurrentSkipListSet`?  It looks there are no concurrent accesses to `ints`

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4SolutionReader.java [hunk 17]
```diff
@@ -354,9 +354,22 @@ private A4SolutionReader(Iterable<Sig> sigs, XMLNode xml) throws IOException, Er
...
             }
+        // Inserting the exceeded integers into the atoms list
+        for (XMLNode i : inst){
+            if (i.getAttribute("label").equals("Int")) {
+                for (XMLNode y: i) {
+                    try{
+                        int value = Integer.valueOf(y.getAttribute("value"));
+                        atoms.add(String.valueOf(value));
+                    }catch (NumberFormatException e) {
+                        // throw new ErrorSyntax("The XML file must containt integer type instead of:" + y.getAttribute("value"));
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4SolutionReader.java]: please handle this exception (same comment as for ignoring `NumberFormatException` about

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java [hunk 18]
```diff
@@ -90,22 +113,55 @@ private TupleSet computeLowerBound(List<Tuple> atoms, final PrimSig sig) throws
...
-            for (n = n - upper.size(); n > 0; n--) {
-                Tuple atom = atoms.remove(atoms.size() - 1);
-                // If MUST<SCOPE and s is exact, then add fresh atoms to both
-                // LOWERBOUND and UPPERBOUND.
-                // If MUST<SCOPE and s is inexact but toplevel, then add fresh
-                // atoms to the UPPERBOUND.
-                if (isExact)
+        boolean hasLower = sc.hasLower(sig);
+
+        if (hasLower)
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java]: please add curly braces around the `then` and `else` clauses of this `if`

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java [hunk 19]
```diff
@@ -198,6 +254,15 @@ private Expression allocateSubsetSig(SubsetSig sig) throws Err {
             return sum;
         }
         // Allocate a relation for this subset sig, then bound it
+        String label = (sig.label.startsWith("this/") ?
+                sig.label.substring(5):
+                    sig.label)+"%";
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java]: this is a very strange formatting of the ternary operator.  I suggest putting everything in a single line (and adding spaces around both `?` and `:`

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java [hunk 20]
```diff
@@ -376,31 +441,76 @@ private BoundsComputer(A4Reporter rep, A4Solution sol, ScopeComputer sc, Iterabl
...
+                        //TODO: change the ub to lower bound and ub2 to upper bound 
+                        r = sol.addRel(s.label+"."+f.label, lb, ub2);
+                        sol.addField(f, isOne ? sol.a2k(s).product(r) : r);
                     }
-                    ub.addAll(upper);
+                }else{
+                    for(List<PrimSig> p:t.fold()) {
+                        TupleSet upper=null;
+                        for(PrimSig b:p) {
+
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java]: empty line after open curly brace

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/TranslateAlloyToKodkod.java [hunk 21]
```diff
@@ -202,6 +203,14 @@ private Expression a2k(Sig x) throws Err {
             return frame.a2k(x);
     }
 
+    /** Returns the expression corresponding to the given Bounds. */
+    private Expression a2k(Bounds x)     throws Err {
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/TranslateAlloyToKodkod.java]: too many consecutive spaces

File: org.alloytools.alloy.core/src/test/java/org/alloytools/alloy/core/AlloyModelsPartialInstancesTest.java [hunk 22]
```diff
@@ -0,0 +1,110 @@
+package org.alloytools.alloy.core;
+
+
```
Comment (aleksandarmilicevic) [inline org.alloytools.alloy.core/src/test/java/org/alloytools/alloy/core/AlloyModelsPartialInstancesTest.java]: too many empty lines

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java (hunk 1 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: removed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java (hunk 2 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: added here and other files.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java (hunk 3 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java (hunk 4 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java (hunk 5 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java (hunk 6 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java (hunk 7 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/Bounds.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java (hunk 8 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java (hunk 9 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: updated!

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java (hunk 10 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java (hunk 11 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java (hunk 12 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java (hunk 13 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/ast/CommandScope.java]: fixed

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompSym.java (hunk 14 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompSym.java]: removed this file and other generated files. The `.gitingore` file is also updated.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java (hunk 15 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java]: it's for filtering integers. non-integers could be safely ignored.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java (hunk 16 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4Solution.java]: good point! Changed to TreeSet.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4SolutionReader.java (hunk 17 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/A4SolutionReader.java]: Good catch! fixed.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java (hunk 18 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java]: fixed.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java (hunk 19 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java]: fixed.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java (hunk 20 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/BoundsComputer.java]: fixed here and the next lines.

File: org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/TranslateAlloyToKodkod.java (hunk 21 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/translator/TranslateAlloyToKodkod.java]: fixed!

File: org.alloytools.alloy.core/src/test/java/org/alloytools/alloy/core/AlloyModelsPartialInstancesTest.java (hunk 22 above)
Comment (vmontagh) [reply org.alloytools.alloy.core/src/test/java/org/alloytools/alloy/core/AlloyModelsPartialInstancesTest.java]: fixed!

**Comment by pkriens** [comment]
Is there a description of the syntax change? Was this change agreed upon with Daniel?

A bit concerned with all the whitespace changes. The way the Eclipse project is setup should make formatting automatically, was this edited in another editor?

**Comment by pkriens** [comment]
I initiated a discussion on the AlloyTools board because this PR causes a change in the syntax of Alloy.

There Daniel indicated that he want to see different proposals to this problem. I distributed the paper this change was based on and someone else submitted a proposal for partial instances on a model that did not require changes to the syntax by recognizing them based on some special patterns.

How are you planning to continue?

**Comment by aleksandarmilicevic** [comment]
> The CompParser, Lexer, etc. file were deleted. This makes it extremely inconvenient to edit the project in Eclipse. We had decided I think that these files were stored in generated form to not force everybody to do a very inconvenient gradle build all the time. Since syntax changes are extremely rare this is is a small price for such a convenience.

Auto-generated files should not be version-controlled.  You don't have to use gradle all the time; you can use it once to generate those files, and from then on you can use Eclipse or whatever other IDE you want.

**Comment by pkriens** [comment]
@aleksandarmilicevic  I recall we had a discussion about this? I thought you accepted this because it is so inconvenient to always have to run gradle. Since syntax changes are very rare you it is easy to get out of sync and get weird errors if you don't put them under source control. 
Surprised, didn't we have this discussion?

File: .gitignore [hunk 23]
```diff
@@ -6,9 +6,9 @@ bin_test/
 target/
 enroute.zip
 .metadata
-org.alloytools.alloy.application/src/main/java/edu/mit/csail/sdg/alloy4compiler/parser/CompLexer.java
-org.alloytools.alloy.application/src/main/java/edu/mit/csail/sdg/alloy4compiler/parser/CompParser.java
-org.alloytools.alloy.application/src/main/java/edu/mit/csail/sdg/alloy4compiler/parser/CompSym.java
+org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompLexer.java
+org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompParser.java
+org.alloytools.alloy.core/src/main/java/edu/mit/csail/sdg/parser/CompSym.java
```
Comment (aleksandarmilicevic) [inline .gitignore]: why did this change?

**Comment by dnjackson** [comment]
Thanks for all your good work on this, Vajih!

This is actually the first proposal for a major change to Alloy, so we've had to figure out our process. See below for what we have in mind.

Our current concern is that it seems that are still regressions in which your version produces different results for existing models.

--Daniel

Process for Contributions to Alloy Codebase or Language

We welcome contributions to the Alloy toolkit and proposed language changes. Here is the process we are currently using for incorporating such contributions; post a message to the Alloy Google Forum to suggest an improvement.

1. If the contribution is a bug fix or improvement in performance that does not require a change to the documentation, you can just submit a pull request in the Alloy repo which will be considered by the Alloy maintainers. Please make sure to include test cases to ensure there are no regressions, and documentation explaining what you changed and why.

2. If the contribution is a major new piece of functionality, especially if it involves changes or extensions to the Alloy language:

— Submit a pull request that includes test cases and documentation explaining the contribution. In the case of a change that affects users, you should include a short FAQ and brief documentation to be included in the language description (eg, grammar rules, semantics). You should also include executables and sample models so other members of the community can experiment with your new version. Contact Peter Kriens if you need help building the executables. Most importantly, you should include a document that gives a brief outline of the proposal explaining it to the members of the Alloy forum. This document can be in any format you choose; it could be a short slide deck, or a technical memo.

— The Alloy maintainers will announce the proposal to the forum, and after some period of discussion, the Alloy Board will meet to consider the proposal. If accepted, the proposed changes will be included in the official release. If not, you may be asked to alter your proposal. Note also that in cases where your proposal involves a feature that others have also proposed, the Board may want to group several proposals together, which may result in things taking a bit longer.

**Comment by vmontagh** [comment]
Thanks for reviewing and explaining the contribution process. I have seen others interest in the work and decided to submit this PR showing the idea and implementation. What proposed here was mostly done years ago for research experiments ran by colleagues and myself. I also think releasing a stable implementation of this feature needs more effort. Regardless of the mentioned regressions, I believe the PR offers enough to clarify the new syntax and a way of realization. I'll be glad to improve the implementation and mitigate regressions if there is a consensus on a future for this PR.

**Comment by pkriens** [comment]
This is outdated. If still interested in supoorting this please update this for Alloy 6

**Pull request closed without merging**